onnx/defs/nn/defs.cc

```diff
@@ -2074,4 +2074,37 @@ ONNX_OPERATOR_SET_SCHEMA(
              {{"Processed_STD"}, "Add", {"STD", "Epsilon"}},
              {{"Y"}, "Div", {"X_variance", "Processed_STD"}}})));
 
+static const char* celu_ver11_doc = R"DOC(
+       Continuously Differentiable Exponential Linear Units:
+       Perform the linear unit element-wise on the input tensor X
+       using formula: <br/> ``` max(0,x) + min(0,α*(exp(x/α)−1)) ```
+)DOC";
+
+static float celu_default_alpha = 1.0;
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Celu,
+    11,
+    OpSchema()
+       .SetDoc(celu_ver11_doc)
+       .Input(0, "X", "Input tensor", "T")
+       .Output(0, "Y", "Output tensor", "T")
+       .Attr(
+           "alpha",
+           "The Alpha value in Celu formula which control the shape of "
+           "the unit. The default value is 1.0.",
+           AttributeProto::FLOAT,
+           celu_default_alpha)
+       .TypeConstraint(
+           "T",
+           {"tensor(float16)", "tensor(float)", "tensor(double)"},
+           "Constrain input and output types to all numeric tensors.")
+       .FunctionBody(FunctionBodyHelper::BuildNodes(
+           {// nodes: {outputs, op, inputs, attributes}
+            {{"X_alpha"},
+             "Div",
+             {"X", "alpha"}
```

Comment:
I can't figure out how to reference the attribute of the `Celu` instruction as an argument of the node `Div`.

I had a look at the helpers `FunctionBodyHelper::BuildNodes`, `FunctionBodyHelper::Const` and `MakeRefAttribute` without success.

Could you show me some documentation / example of this usage?

Reply:
There is a comment about this on `FunctionBodyHelper::BuildNodes` [here](https://github.com/onnx/onnx/blob/121114107c38ec49bab4955c13cb82b7675bfc47/onnx/defs/function.h#L79-L80), and it's used in the `MeanVarianceNormalization` operator function [here](https://github.com/onnx/onnx/blob/121114107c38ec49bab4955c13cb82b7675bfc47/onnx/defs/nn/defs.cc#L2057-L2075). Shout if you have trouble :)

Reply:
Thanks you for your answer. :)

Unfortunately I did read this line and the usage in MeanVarianceNormalization but I am still confused. I tried different syntax that compile, but since the shape inference test fail it is probably not right graph.

As I understand, I can create a graph equivalent to `Div(X, alpha=alpha)` using
```
           {{"X_alpha"},
             "Div",
             {"X"},
             {MakeRefAttribute("alpha", AttributeProto::FLOAT)}
            },
```
But it doesn't seams to be what I am lloking for, probably because `Div` have two input and 0 attribute.

How can write the equivalent of `Div(X, alpha)` (i.e. use this reference as the second argument of Div)? I would like to write
```
           {{"X_alpha"},
             "Div",
             {"X", MakeRefAttribute("alpha", AttributeProto::FLOAT)}
            }
```
but obviously this is not possible since `std::String != AttributeProtoWrapper` 😅

In the MeanVarianceNormalization it seams all the usage of axis simply forward the attribute to the underlying Ops, right ?

Reply:
Thanks you for your answer. :)

Unfortunately I did read this line and the usage in MeanVarianceNormalization but I am still confused. I tried different syntax that compile, but since the shape inference test fail it is probably not right graph.

As I understand, I can create a graph equivalent to `Div(X, alpha=alpha)` using
```
           {{"X_alpha"},
             "Div",
             {"X"},
             {MakeRefAttribute("alpha", AttributeProto::FLOAT)}
            },
```
But it doesn't seams to be what I am lloking for, probably because `Div` have two input and 0 attribute.

How can write the equivalent of `Div(X, alpha)` (i.e. use this reference as the second argument of Div)? I would like to write
```
           {{"X_alpha"},
             "Div",
             {"X", MakeRefAttribute("alpha", AttributeProto::FLOAT)}
            }
```
but obviously this is not possible since `std::String != AttributeProtoWrapper` 😅

In the MeanVarianceNormalization it seams all the usage of axis simply forward the attribute to the underlying Ops, right ?

Reply:
Ah okay, I see, you are quite right. I think you should be able to move from a reference to a value by adding a `Constant` node.
```
FunctionBodyHelper::BuildNodes(
           {// nodes: {outputs, op, inputs, attributes}
            {{"alpha"}, "Constant", {}, {MakeRefAttribute("alpha", AttributeProto::FLOAT)}},
            {{"X_alpha"}, "Div", {"X", "alpha"}, {}},
            {{"Y"}, "Elu", {"X_alpha"}}})
```